### ISSUE TITLE
Microsoft.Azure.Devices.Client-1.29.1-preview-001 multiple disconnects when using MQTT with device streaming APIs

### ISSUE THREAD

**Issue by Harsh-m-Mehta** [Azure/azure-iot-sdk-csharp#1636]
We are using 1.29.1-preview-001 SDK for our project. We started using this SDK to enable device-stream feature. After updating SDK, parallel operations are not happening. For example, if one direct method call is going on, another direct method call is reaching to device once current running direct method call finishes its execution. Also if one direct method call is going on then device twin update and message to device reaches to device once execution of current running direct method.
This is not the case with non device-streams SDKs, earlier switching to this SDK I was using 1.25.0 version of SDK, where these functionalities were working fine. We are using it over AMQP.

Please help.

Thanks,
Harsh Mehta 

# Context

- **OS, version, SKU and CPU architecture used:** <VERSION> (Windows 10 Desktop x64, Ubuntu 15.04 x86, Windows 10 IoT Core arm32, etc.)
- **Application's .NET Target Framework :** <NETVERSION> (See https://docs.microsoft.com/en-us/dotnet/standard/frameworks. E.g. netcoreapp2.1, net451, uap10.0, xamarin)
- **Device:** <DEVICETYPE> (Laptop, Raspberry PI3, Android APIv25 etc.)
- **SDK version used:** <VERSION> (Please include the NuGet package version for all involved components)

## Description of the issue
Please be as detailed as possible: which feature has a problem, how often does it fail, etc.

## Code sample exhibiting the issue
Please remove any connection string information!

## Console log of the issue
Consider setting the DEBUG environment variable to '*'. This will produce a much more verbose output that will help debugging
Don't forget to remove any connection string information!

**Comment by azabbasi** [comment]
@Harsh-m-Mehta Thank you for bringing this up to us. 
Can you share some code snippets on how you set up your device client and elaborate on how you identified the issue?

**Comment by Harsh-m-Mehta** [comment]
@azabbasi Thank you for your prompt response. Below is the code section where I create the device client.
private DeviceClient _deviceClient = null;

private async Task ConnectDevice()
{
            try
            {
                deviceStatus = false;
                //Checking device status and dispose it if not null
                try
                {
                    if (_deviceClient != null)
                    {
                        _deviceClient.Dispose();
                        _deviceClient = null;
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine(DateTime.UtcNow.ToString() + " : Failed to dispose device client object " + ex.ToString());
                }
                //ConsoleEventListener _listener = new ConsoleEventListener("Microsoft-Azure-Devices-Device-Client", DeviceId + ".txt");
                //Device connectivity
                Console.WriteLine(DateTime.UtcNow.ToString() + " : Creating Device Client");
                Console.WriteLine(DateTime.UtcNow.ToString() + " : Connection string is {0}", _deviceConfig.ConnectionString);

                Console.WriteLine(DateTime.UtcNow.ToString() + " : serial number {0}", _deviceConfig.SerialNumber);
                _deviceClient = DeviceClient.CreateFromConnectionString(_deviceConfig.ConnectionString, TransportType.Amqp);
                //_deviceClient.SetRetryPolicy(new CustomRetryPolicy());
                serialNumber = _deviceConfig.SerialNumber;

                if (_deviceClient == null)
                {
                    Console.WriteLine(DateTime.UtcNow.ToString() + " : Failed to create DeviceClient");
                    return;
                }

#region Reading Device Twin Property for Certificate and Device Host Name
		        try
		        {
                    //Read device twin for .pem certificate and for GatewayHostName
                    Task<Twin> _deviceTwin = _deviceClient.GetTwinAsync();
		            if (_deviceTwin != null)
		            {
		                JObject Objtwin = JObject.Parse(_deviceTwin.Result.ToJson());
                    	//Checking for GatewayHostName
                    	if (Objtwin["properties"]["desired"]["GatewayHostName"] != null)
                    	{
                            if (!string.IsNullOrEmpty(Convert.ToString(Objtwin["properties"]["desired"]["GatewayHostName"])))
                            {
                            	//Installation of a certificate
                            	InstallCertificate(Convert.ToString(Objtwin["properties"]["desired"]["TlsCertsLink"]));
                            	//Connect deviceclient again, with GatewayHostName
                            	await ReconnectViaGateway(Convert.ToString(Objtwin["properties"]["desired"]["GatewayHostName"]));
                            	return;
                            }
                    	}
		            }
		        }
		        catch(Exception e)
		        {
			        Console.WriteLine("Exception while reading device twin for GatewayHostName");
		        }

                #endregion  

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Device Client Created");
}

And this is how I call ConnectDevice function.
await ConnectDevice().ConfigureAwait(false);

In our application a point come where we want to execute two directmethods in parallel, at that point of time we observed that even if we executed second direct method while first direct method execution is going on, it reaches to the device once first direct method completes its execution. Then I checked this thing with scenario where my long running direct method is in execution and if I set device-twin, that also reaches to device once execution of direct method completes. The same case with message to device.
Thanks,
Harsh Mehta

**Comment by azabbasi** [comment]
Something that I have noticed right off the bat is that you are not awaiting the `_deviceTwin` task 

in your try block you have 
`Task<Twin> _deviceTwin = _deviceClient.GetTwinASync();`

This task needs to be awaited before the Results can be used in the following `if` statement.

**Comment by azabbasi** [comment]
As far as the issue with calling ConnectDevice() in parallel goes, Let's take a look at the piece of code you have got that is running them in parallel so we can inspect and see if there are any irregularities there as well.

**Comment by Harsh-m-Mehta** [comment]
Hi @azabbasi thanks for your response. We are using Task<Twin> _deviceTwin = _deviceClient.GetTwinAsync();, to understand whether device is associated with gateway(IoT edge device) or not. So based on that we can update connection string with gateway host name. Below is the code snippet where we attach the directmethod and reported properties handlers.

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Registering Heartbeat Timer");

                this.heartbeatTimer = new Timer(30000);
                this.heartbeatTimer.Elapsed += new ElapsedEventHandler(SendHeartbeatAsync);
                this.heartbeatTimer.AutoReset = true;
                this.heartbeatTimer.Start();

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Registered Heartbeat Timer");

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attaching Connection Status Changes Handler");

                _deviceClient.SetConnectionStatusChangesHandler(CallBackForConnectionStatusChangesHandler);

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attached Connection Status Changes Handler");

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attaching Direct Method Handlers");

                string methodType = string.Empty;

		try
		{

                    //Register for Direct Method Callback
                    foreach (MethodInfo method in _deviceTypeConfig.DirectMethods)
                    {
                        object[] attrib = method.GetCustomAttributes(typeof(CallBackMethodAttribute), true);

                        foreach (Object attribute in attrib)
                        {
                            CallBackMethodAttribute a = (CallBackMethodAttribute)attribute;
                            methodType = a.MethodType;
                        }

                        if (method.ReturnType == typeof(Task<MethodResponse>) && methodType.ToLower().Equals("directmethod"))
                        {
                            var handler = method.CreateDelegate(typeof(MethodCallback), this) as MethodCallback;
                            await _deviceClient.SetMethodHandlerAsync(method.Name, handler, null);
                        }   
                    }   

                    Console.WriteLine(DateTime.UtcNow.ToString() + " : Attached Direct Method Handlers");
		}
		catch(Exception directMethod)
                {
                    Connect_Flag = 1;
                    Console.WriteLine(DateTime.UtcNow.ToString() + " : Exception while Attaching Direct Method Callback : " + directMethod.ToString());
                }

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attaching Default Method Callback");

                await _deviceClient.SetMethodDefaultHandlerAsync(DefaultDirectMethod, null);

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attached Default Method Callback");

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attaching Desired Properties Callback");

                **//Register for Desired Properties Callback
                await _deviceClient.SetDesiredPropertyUpdateCallbackAsync(OnDesiredPropertyChanged, null).ConfigureAwait(false);**

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attached Desired Properties Callback");

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Sending Startup Time as Reported Property");

                //Report DateTimeLastDeviceLaunch
                TwinCollection reportedProperties = new TwinCollection();
                reportedProperties["StartUpTime"] = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffff");
                await _deviceClient.UpdateReportedPropertiesAsync(reportedProperties).ConfigureAwait(false);

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Sent Startup Time as Reported Property");

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attaching Cloud to Device Method Handlers");

                //Register for Cloud to Device Messaging to recevive Commands
                ReceiveCommands().ConfigureAwait(false);

                Console.WriteLine(DateTime.UtcNow.ToString() + " : Attached Cloud to Device Method Handlers");


                Console.WriteLine(DateTime.UtcNow.ToString() + " : Device Initialization completed for " + DeviceId);

                deviceStatus = true;
                deviceConnectivity = true;
                isIoTConnectionClosed = false;
            }
            catch (Exception ex)
            {
                Console.WriteLine(DateTime.UtcNow.ToString() + " : Exception at ConnectDevice Method : " + ex.ToString());
            }

The only difference in my working code and none working code is device client SDK. Earlier we are using 1.25.0 version of device client SDK and now to support device streams we switched to 1.29.1-preview-001 SDK. With 1.25.0 we were using MQTT and for this preview version SDK we are using AMQP as transport type.

**Comment by Harsh-m-Mehta** [comment]
Hi @azabbasi  I did some testing and I found that in normal SDKs also if we use AMQP the parallel stuff is not working and if we use MQTT parallel stuff is working fine.

**Comment by Harsh-m-Mehta** [comment]
Parallel operations with AMQP works fine now. We were using Thread.Sleep() call which was causing the issues with AMQP protocol, we replaced the Thread.Sleep() with Task.Delay() and parallel operations works for me.

**Comment by azabbasi** [comment]
@Harsh-m-Mehta 
I have looked into the behavior inconsistency regarding the Method invocation callback between AMQP and MQTT protocols.
With MQTT, the Device Client SDK spins off a new thread to handle each individual method callback so even if you block the thread that is handed to the callback method, the other method callbacks will be executed independently. Whereas with AMQP protocol, The method callback is invoked on the same thread asynchronously so if the user doesn't implement correct asynchronous patterns and block the thread, all the rest of the method invocations will be held off until the thread is released. 
When you use `await Task.Delay(int)` you are setting a bookmark from where the Delay is being called and release the thread to the thread pool so it can pick up other tasks (method callbacks) and execute them asynchronously. if you use `Thread.Sleep(int)` or any other operation that holds on to the thread, you will prevent it from being released.
While this is an inconsistent behavior between the two protocols, if you use correct async patterns in your method callback, you shouldn't have any issues.

**Comment by Harsh-m-Mehta** [comment]
Hi @azabbasi ,
Thanks for the update. Yes I did modify my code to use await Task.Delay(int) then Thread.Sleep(int) with AMQP to make it work.

**Comment by timtay-microsoft** [comment]
@azabbasi @Harsh-m-Mehta is there anything pending on this issue? Seems like everyone is unblocked here.

**Comment by abhipsaMisra** [comment]
We have this issue open to figure out the outstanding issue where @Harsh-m-Mehta is seeing multiple disconnects when using MQTT with device streaming APIs (https://github.com/Azure/azure-iot-sdk-csharp/issues/1636#issuecomment-722187137). A simple device streaming sample with minimal overhead code has been shared, we are waiting for the test results.

We should probably rename the title to highlight the new issue under investigation, or open another issue to track this.

**Comment by azabbasi** [comment]
I have updated the Title so it represent the current issue at hand.
We are still in communication with the user over Email.

**Comment by azabbasi** [comment]
I am closing this issue since there has been no response. 
@Harsh-m-Mehta please feel free to reopen the issue if you are still blocked on it.

**Comment by az-iot-builder-01** [comment]
@azabbasi, @Harsh-m-Mehta, @timtay-microsoft, @abhipsaMisra, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
